torch/fx/experimental/symbolic_shapes.py

```diff
@@ -916,9 +942,258 @@ def _print_Symbol(self, expr) -> str:
+        assert not self.is_source_registered(source)
+        self.register_source(source, name=name, type=type)
+        self.parameter_infos.append(self.get_source_info(source))
+
+    def register_unique_name(self, source: Union[str, Source]) -> str:
+        from torch._dynamo.source import TensorPropertySource
+
+        if isinstance(source, TensorPropertySource):
+            source_name = f"{self.source_expr_to_info[source.base.name()].name}_{source.prop.name.lower()}"
+        elif isinstance(source, Source):
+            source_name = source.name()
+        else:
+            source_name = source
+
+        return self.unique_name_set.mangle_and_add(source_name)
+
+        # Makes sure there are no repeated names.
+        # It's very unlikely the loop-body will be executed more than once.
+        counter = 1
+        unique_source_name = source_name
+        while unique_source_name in self._unique_names:
+            unique_source_name = source_name + str(counter)
+            counter += 1
+
+        self._unique_names.add(unique_source_name)
+        return unique_source_name
+
+    def flat_source_ref(self, source: Source) -> str:
+        from torch._dynamo.source import TensorPropertySource
+
+        # We need to call 'source_ref' here due to its side effects.
+        ref = self.source_ref(source)
+
+        info = self.get_source_info(source)
+        index_expr = ""
+        if isinstance(source, TensorPropertySource) and source.idx is not None:
+            index_expr = f"[{source.idx}]"
+        return f"{info.name}{index_expr}"
+
+    def compile(self, exprs: List[str], python_exprs: List[str]) -> "GuardCompiler.SymbolicShapesExpr":
+        """Generates and compiles the guard function in C++.
+
+        1. Generates a C++ function based on its object's registered parameters
+        2. Compiles the C++ function, and loads it with the help of `CppCodeCache`
+
+        The overall structure of the generated function is as follows:
+
+        extern "C" int guard(PyObject** inputs) {
+            // Inputs unpacking
+            // Source creation (e.g. sizes, strides, and storage offset variables are defined)
+            // Return the actual guard expression
+        }
+        """
+        from torch._dynamo.source import TensorPropertySource, TensorProperty
+
+        if len(exprs) == 0:
+            return self.SymbolicShapesExpr([], "True")
+
+        fn_name = "___symbolic_shape_fn"
+
+        # 1. Function call
+        #     a) Get all the argument expressions for each parameter
+        call_arguments = ", ".join(info.source.name() for info in self.parameter_infos)
+        #     b) Generate the actual function call
+        call_expression = f"{fn_name}({call_arguments})"
+
+        # 2. Function definition
+        #     a) Generate the actual guard expression
+        guard_expression = " && ".join(exprs)
```

Comment:
Another thing that might be worth trying: 
```suggestion
        guard_expression = " & ".join(exprs)
```

With short-circuiting you're effectively creating an if-else for every guard, but the bitwise operation is branch-less so the code is simpler. Though this would mean failed guard checks would be more expensive (though at least consistent).

Reply:
Agreed. I will change this and see how that works out.